rust/dialog-common-macros/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+/// 2. `cargo test --target wasm32-unknown-unknown` - Unit tests run in wasm
+/// 3. `RUSTFLAGS="--cfg dialog_test_wasm" cargo test` - Integration tests run in wasm
+///    (native provider spawns wasm inner tests)
+///
+/// # Usage
+///
+/// ## Unit tests
+///
+/// Tests that do not require external services:
+///
+/// ```rs
+/// // Sync test
+/// #[dialog_common::test]
+/// fn it_works() {
+///     assert_eq!(2 + 2, 4);
+/// }
+///
+/// // Async test
+/// #[dialog_common::test]
+/// async fn it_works_async() -> anyhow::Result<()> {
+///     assert_eq!(2 + 2, 4);
+///     Ok(())
+/// }
+/// ```
+///
+/// Unit tests are gated with `#[cfg(not(dialog_test_wasm))]` so they don't
+/// run during wasm integration test runs (case 3 above).
+///
+/// ## Integration tests
+///
+/// Tests that need external services (S3, databases, servers, etc.):
+///
+/// ### Step 1: Define the address (what the test receives)
+///
+/// The address is serializable so it can be passed to wasm tests via environment:
+///
+/// ```rs
+/// use serde::{Deserialize, Serialize};
+///
+/// #[derive(Debug, Clone, Serialize, Deserialize)]
+/// pub struct ServerAddress {
+///     pub endpoint: String,
+///     pub port: u16,
+/// }
+/// ```
+///
+/// ### Step 2: Define the provider (native only)
+///
+/// Use `#[dialog_common::provider]` to create a function that starts the service
+/// and returns its address:
+///
+/// ```rs
+/// use dialog_common::helpers::{Provider, Service};
+///
+/// pub struct LocalServer { /* ... */ }
+///
+/// impl Provider for LocalServer {
+///     async fn stop(self) -> anyhow::Result<()> {
+///         Ok(())
+///     }
+/// }
+///
+/// #[dialog_common::provider]
+/// async fn server(_settings: ()) -> anyhow::Result<Service<ServerAddress, LocalServer>> {
+///     let server = LocalServer { /* start server */ };
+///     Ok(Service::new(
+///         ServerAddress { endpoint: "...".into(), port: 8080 },
+///         server
+///     ))
+/// }
+/// ```
+///
+/// ### Step 3: Write the test
+///
+/// When a function takes an address parameter, the macro generates:
+///
+/// - **Native inline provider**: Starts service, runs test via `tokio::spawn`, stops service
+/// - **Native wasm-spawn provider**: Starts service, spawns `cargo test --target wasm32...`
+/// - **Wasm inner**: Deserializes address from env var, runs test
+///
+/// ```rs
+/// #[dialog_common::test]
+/// async fn it_connects_to_server(server: ServerAddress) -> anyhow::Result<()> {
+///     // Use server.endpoint to connect...
+///     Ok(())
+/// }
+/// ```
+///
+/// ### With custom settings
+///
+/// Provider settings can be customized via macro attributes:
+///
+/// ```rs
+/// #[dialog_common::test(port = 9000, tls = true)]
+/// async fn it_uses_custom_port(server: ServerAddress) -> anyhow::Result<()> {
+///     assert_eq!(server.port, 9000);
+///     Ok(())
+/// }
+/// ```
+#[proc_macro_attribute]
+pub fn test(attr: TokenStream, item: TokenStream) -> TokenStream {
```

Comment:
It's a nuanced thing, but I think it would be better to call this a `integration_test`. It makes it immediately clear why this is different from other tests (and implies that it is a most costly test to add, which it is). Also: we may want to add a `test` macro at some point to spit out all the boilerplate that we currently add to most unit tests, and at that point the distinction will be important.